### PR TITLE
feat(rust): generalize the cloud storage builders

### PR DESCRIPTION
**Pull request by winding-lines**
This PR generalizes the parquet support for cloud urls. It enables all the backends supported by `object_store`.  

This PR threads through the code the required cloud options, as opposed to using the process environment. This reduces the magic provided by global state (std::env) at the expense of more changes for the different layers. I personally prefer more explicit options/settings passing. 

This PR also breaks the existing `object_store.rs` into a module, that file was getting bigger and had a poor cohesion.

Feedback appreciated.

### CODE REVIEW

**Comment by winding-lines** [comment]
I got some help from the developers of `object-store` crate, they provided an API that is nicer for us to use. Many thanks :)

This code is ready to review at your convenience. I am not sure how to fully enable it in the Python side... @ritchie46

File: examples/read_parquet_cloud/src/main.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
-use std::env;
+use std::collections::HashMap;
```
Comment (ritchie46) [inline examples/read_parquet_cloud/src/main.rs:1]: Can you use `PlHashMap`? That's the hashmap we use within polars.

File: polars/polars-io/Cargo.toml [hunk 2]
```diff
@@ -51,7 +54,7 @@ lexical-core = { version = "0.8", optional = true }
 memchr = "2.5"
 memmap = { package = "memmap2", version = "0.5.2", optional = true }
 num.workspace = true
-object_store = { version = "0.5.2", features = ["aws"], optional = true }
+object_store.workspace = true
```
Comment (ritchie46) [inline polars/polars-io/Cargo.toml:57]: Is this still optional like this?

**Review by ritchie46 — COMMENTED** [review]
Thanks @winding-lines! I have a few small remarks.

File: examples/read_parquet_cloud/src/main.rs (hunk 1 above)
Comment (winding-lines) [reply examples/read_parquet_cloud/src/main.rs:1]: @ritchie46 sorry about this the interface to `object_store` uses a hash map. Let me see if I can use a different API on their side, if not I can use a PLHashMap and translate it just before the call.

Comment (ritchie46) [reply examples/read_parquet_cloud/src/main.rs:1]: oh, in that case nevermind.

Comment (winding-lines) [reply examples/read_parquet_cloud/src/main.rs]: Ugh, I was totally wrong and you were right to push on this. The initial version of the API did use a HashMap but in the review process they removed it. I adapted the improved API and also removed the HashMap.

File: polars/polars-io/Cargo.toml (hunk 2 above)
Comment (winding-lines) [reply polars/polars-io/Cargo.toml]: I cannot figure the answer from the docs. I moved to explicitly defining the dependencies.

**Comment by winding-lines** [comment]
@ritchie46 addressed feedback, tests are passing, ready for more feedback or merge 🤗

**Comment by chitralverma** [comment]
@winding-lines when object_store adds more backends in the future, will it require changes in polars as well. because I see that in their main/ master branch HTTP has also been added.

**Comment by winding-lines** [comment]
@chitralverma in the current iteration the deltalake and object_store teams have refactored some code out of the former and put in the latter. I think we can also contribute our current layer upstream so that any future changes will be integrated just by recompiling.

Given that this PR has been open for a while my preference would be to merge it and then release a polars version so that I can do more testing at work.

Let me know what you think :)

File: polars/polars-io/Cargo.toml [hunk 3]
```diff
@@ -51,7 +54,7 @@ lexical-core = { version = "0.8", optional = true }
 memchr = "2.5"
 memmap = { package = "memmap2", version = "0.5.2", optional = true }
 num.workspace = true
-object_store = { version = "0.5.2", features = ["aws"], optional = true }
+object_store = { git = "https://github.com/apache/arrow-rs", default-features = false, rev = "c74665808439cb7020fb1cfb74b376a136c73259" }
```
Comment (ritchie46) [inline polars/polars-io/Cargo.toml:57]: Can you make this optional?

File: polars/polars-core/Cargo.toml [hunk 4]
```diff
@@ -161,6 +167,7 @@ hex = { version = "0.4", optional = true }
 indexmap = { version = "1", features = ["std"] }
 ndarray = { version = "0.15", optional = true, default_features = false }
 num.workspace = true
+object_store = { git = "https://github.com/apache/arrow-rs", default-features = false, rev = "c74665808439cb7020fb1cfb74b376a136c73259" }
```
Comment (ritchie46) [inline polars/polars-core/Cargo.toml:170]: Can you make this optional?

**Review by ritchie46 — COMMENTED** [review]
Thanks a lot @winding-lines. I have only two remarks with respect to dependencies. 

If you can fix those and the conflicts it is good to go.

File: polars/polars-io/Cargo.toml (hunk 3 above)
Comment (winding-lines) [reply polars/polars-io/Cargo.toml]: Ugh, sorry. Made optional and switched to use the published version of the crate.

File: polars/polars-core/Cargo.toml (hunk 4 above)
Comment (winding-lines) [reply polars/polars-core/Cargo.toml]: Done.

**Comment by winding-lines** [comment]
@ritchie46 addressed feedback, tests are passing :-) 

Many thanks!

**Comment by winding-lines** [comment]
thanks @ritchie46 ❤️

**Pull request merged**